File: app/src/main/java/vedam/subkuch/ui/directory/DirectoryFragment.java
package vedam.subkuch.ui.directory;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import com.android.volley.Response;

import java.util.ArrayList;

import vedam.subkuch.R;
import vedam.subkuch.base.BaseListFragment;
import vedam.subkuch.helpers.Constants;
import vedam.subkuch.network.DataFetcher;
import vedam.subkuch.ui.directory.models.Category;
import vedam.subkuch.ui.directory.models.CategoryResponse;
import vedam.subkuch.ui.stafftrack.StaffTrackActivity;
import vedam.subkuch.utils.UiUtil;

public class DirectoryFragment extends BaseListFragment {

    private ArrayList<Category> categories;

    public DirectoryFragment() {
        // Required empty public constructor
    }

    public static DirectoryFragment newInstance() {
        return new DirectoryFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_directory, container, false);
    }

    public void onViewCreated(@NonNull View v, Bundle savedInstanceState) {
        super.onViewCreated(v, savedInstanceState);
        getCategories();
    }

    private void getCategories() {
        UiUtil.showProgressDialog(context, R.string.please_wait);
        DataFetcher.getCategories(context, onCategorySuccessListener, CategoryResponse.class, onErrorListener);

    }

    private Response.Listener<CategoryResponse> onCategorySuccessListener = response -> {

        UiUtil.cancelProgressDialog();
        if (response != null && response.getStatus().equals(Constants.TRUE)) {
            categories = response.getCategoryResult().getCategories();
            loadValues();
        } else
            UiUtil.showToast(context, getString(R.string.no_data));
    };

    private void loadValues() {

        ArrayAdapter<Category> adapter = new ArrayAdapter<>(context, android.R.layout.simple_list_item_1,
                android.R.id.text1, categories);
        setListAdapter(adapter);
    }

    @Override
    public void onListItemClick(ListView l, View v, int position, long id) {
        super.onListItemClick(l, v, position, id);


        Intent intent = new Intent(getActivity(),
                SubDirectoryActivity.class);
        Category category = categories.get(position);
        intent.putExtra(Constants.EXTRA_CATEGORY_ID,
                category.getCategoryId());
        intent.putExtra(Constants.EXTRA_CATEGORY_NAME, category.getName());
        startActivity(intent);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        menu.clear();
        inflater.inflate(R.menu.directory, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {
            case R.id.action_add:
                startActivity(new Intent(getActivity(), AddDirectoryActivity.class));
                return true;
            case R.id.action_staff:
                startActivity(new Intent(getActivity(), StaffTrackActivity.class));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

File: app/src/main/java/vedam/subkuch/ui/directory/DirectoryFragment.kt
package vedam.subkuch.ui.directory

import android.content.Intent
import android.os.Bundle
import android.view.*
import android.widget.ArrayAdapter
import android.widget.ListView
import androidx.lifecycle.lifecycleScope
import com.android.volley.Response
import com.google.firebase.crashlytics.FirebaseCrashlytics
import kotlinx.coroutines.flow.*
import vedam.subkuch.R
import vedam.subkuch.base.BaseListFragment
import vedam.subkuch.databinding.FragmentDirectoryBinding
import vedam.subkuch.helpers.Constants
import vedam.subkuch.network.DataFetcher.getCategories
import vedam.subkuch.network.DataFetcher.searchBusiness
import vedam.subkuch.ui.directory.models.Business
import vedam.subkuch.ui.directory.models.Category
import vedam.subkuch.ui.directory.models.CategoryResponse
import vedam.subkuch.ui.directory.models.DirectoryDetailResponse
import vedam.subkuch.ui.shopping.show
import vedam.subkuch.utils.Extensions.Companion.hideKeyboard
import vedam.subkuch.utils.Extensions.Companion.textChanges
import vedam.subkuch.utils.UiUtil
import java.util.*

class DirectoryFragment : BaseListFragment() {
    private var categories: ArrayList<Category>? = null
    private var binding: FragmentDirectoryBinding? = null
    private var businesses: ArrayList<Business> = arrayListOf()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        setHasOptionsMenu(true)
        binding = FragmentDirectoryBinding.inflate(inflater, container, false)
        // Inflate the layout for this fragment
        return binding?.root
    }

    override fun onViewCreated(v: View, savedInstanceState: Bundle?) {
        super.onViewCreated(v, savedInstanceState)
        getCategories()
        initUI()
    }

    private fun initUI() {
        binding?.cvSearch?.show()
        binding?.etSearch!!.textChanges()
            .debounce(300)
            .flatMapLatest {
                showAutoSuggestion(it)
            }
            .launchIn(lifecycleScope)

        binding?.etSearch?.setOnItemClickListener { _, _, position, _ ->

            try {

                val business = businesses[position]

                binding?.etSearch?.setText("")
                binding?.etSearch?.dismissDropDown()
                binding?.etSearch?.hideKeyboard()

                addFragmentWithAnimation(
                    R.id.content_frame, DetailFragment.newInstance(business),
                    null, true
                )
            } catch (e: Exception) {
                FirebaseCrashlytics.getInstance().recordException(e)
            }
        }
    }

    private val onSearchSuccessListener = Response.Listener { response: DirectoryDetailResponse? ->
        UiUtil.cancelProgressDialog()
        if (activity != null) if (response != null && response.status == Constants.TRUE) {
            businesses = response.businessesResult.businesses
            val adapter = ArrayAdapter(
                context, android.R.layout.simple_dropdown_item_1line,
                businesses
            )
            binding?.etSearch?.setAdapter(adapter)
            binding?.etSearch?.showDropDown()
//            loadValues(businesses)
        } else UiUtil.showToast(context, getString(R.string.no_data))
    }

    /**
     * Shows auto suggestion to chip input - shows the typed text as suggestion
     * won't make any API call
     * */
    private fun showAutoSuggestion(
        term: CharSequence?
    ): Flow<List<String>> {

        if (term?.trim()?.length ?: 0 > 2)
            searchBusiness(
                context,
                onSearchSuccessListener,
                DirectoryDetailResponse::class.java,
                onErrorListener,
                term
            )
        // returning dummy flow emit
        return flow {
            emit(listOf())
        }
    }

    private fun getCategories() {
        UiUtil.showProgressDialog(context, R.string.please_wait)
        getCategories(
            context,
            onCategorySuccessListener,
            CategoryResponse::class.java,
            onErrorListener
        )
    }

    private val onCategorySuccessListener = Response.Listener { response: CategoryResponse? ->
        UiUtil.cancelProgressDialog()
        if (activity != null) if (response != null && response.status == Constants.TRUE) {
            categories = response.categoryResult.categories
            loadValues()
        } else UiUtil.showToast(context, getString(R.string.no_data))
    }

    private fun loadValues() {
        val adapter = ArrayAdapter(
            context, android.R.layout.simple_list_item_1,
            android.R.id.text1, categories!!
        )
        listAdapter = adapter
    }

    override fun onListItemClick(l: ListView, v: View, position: Int, id: Long) {
        super.onListItemClick(l, v, position, id)
        val intent = Intent(
            activity,
            SubDirectoryActivity::class.java
        )
        val category = categories!![position]
        intent.putExtra(
            Constants.EXTRA_CATEGORY_ID,
            category.categoryId
        )
        intent.putExtra(Constants.EXTRA_CATEGORY_NAME, category.name)
        startActivity(intent)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        menu.clear()
        inflater.inflate(R.menu.directory, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_add) {
            startActivity(Intent(activity, AddDirectoryActivity::class.java))
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        @JvmStatic
        fun newInstance(): DirectoryFragment {
            return DirectoryFragment()
        }
    }
}